### PR TITLE
Added drag and drop functionality to the item ordering of featured articles (Ref #8138)

### PR DESCRIPTION
**Pull request by nishadi**
Pull Request for Issue #8138
#### Summary of Changes

Added ordering Heading button
Added sort-able drag and drop button with drag and drop functionality
Added save ordering functionality to save the ordering when the items are sorted with drag and drop
#### Testing Instructions
1. Go to Back end > Content > Articles
2. Can sort the article items by drag and drop 
3. Go to Back end > Content > Featured Articles
4. Can not sort the article items by drag and drop
5. Apply the patch
6. Go to Back end > Content > Featured Articles
7. Now the view shows the drag and drop functionality

### CODE REVIEW

**Comment by brianteeman** [comment]
This seems to work but it is a little confusing (to me at least) as the number doesnt change on screen until after a page refresh<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

File: administrator/components/com_content/views/featured/tmpl/default.php [hunk 1]
```diff
@@ -23,6 +23,14 @@
...
 $trashed   = $this->state->get('filter.published') == -2 ? true : false;
 $saveOrder = $listOrder == 'fp.ordering';
+
+if ($saveOrder)
+{
+	$saveOrderingUrl = 'index.php?option=com_content&task=featured.saveOrderAjax&tmpl=component';
+	JHtml::_('sortablelist.sortable', 'articleList', 'adminForm', strtolower($listDirn), $saveOrderingUrl);
+}
+
+$assoc = JLanguageAssociations::isEnabled();
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/featured/tmpl/default.php]: hi. why is this added here?

File: administrator/components/com_content/views/featured/tmpl/default.php [hunk 2]
```diff
@@ -23,6 +23,14 @@
 $archived  = $this->state->get('filter.published') == 2 ? true : false;
 $trashed   = $this->state->get('filter.published') == -2 ? true : false;
 $saveOrder = $listOrder == 'fp.ordering';
```
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/featured/tmpl/default.php]: here you say the ordering field is `fp.ordering`.

But, below you use `a.ordering` `asc` as default for the ordering column.
```html
<th width="1%" class="nowrap center hidden-phone">
<?php echo JHtml::_('searchtools.sort', '', 'a.ordering', $listDirn, $listOrder, null, 'asc', 'JGRID_HEADING_ORDERING', 'icon-menu-2'); ?>
</th>
```

And also say the drag is is grouped by category id
```html
sortable-group-id="<?php echo $item->catid; ?>"
```

Also, the search tools filter xml sets the default ordering as `title ASC` in https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/models/forms/filter_featured.xml#L90

Finally the `featured` model sets the default ordering as `title` `asc` in https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/models/featured.php#L201 and in https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/models/featured.php#L218

Can this be normalized?

**Comment by lunalars** [comment]
I think the column "Ordering" should be removed completely (like at articles).
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by nishadi** [comment]
I will update the PR according to the given feedback.

**Comment by JoshuaLewis** [comment]
Subscribed. Let us know when this PR is updated.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by Webdongle** [comment]
I have tested this item :red_circle: unsuccessfully on 8da42a6fa107c58c1a5bb629f6281ac6b961d074<br /><br />Clicking the 'Ordering' column changes the sort by field correctly.  But clicking the 'arrow' column doesn't change the sort by field correctly which prevents the drag/drop from working.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

File: administrator/components/com_content/views/featured/tmpl/default.php (hunk 2 above)
Comment (nishadi) [inline administrator/components/com_content/views/featured/tmpl/default.php]: Misplacing fp.ordering with a.ordering was a mistake since I was referring articles view to get the same view on the featured articles view.

Comment (nishadi) [inline administrator/components/com_content/views/featured/tmpl/default.php]: Can you please explain a bit regarding how to normalize this?

**Comment by nishadi** [comment]
I have updated the PR removing the ordering column as suggested and now the featured article listing looks exactly like the article listing.
Need clarifications on updating and normalization of the default ordering in featured articles.

Thanks

**Comment by Webdongle** [comment]
I have tested this item :white_check_mark: successfully on 1c6d138cfd9cc933b249d1843d044227acf771d2<br /><br />Excellent.  Acts exactly the same as the Articles page.  Front end Featured Articles displays according to order in the featured drag/drop screen.

Full Success<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by lunalars** [comment]
I have tested this item :white_check_mark: successfully on 1c6d138cfd9cc933b249d1843d044227acf771d2<br /><br />Yup, as webdongle said.
Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@nishadi it seems fine now. thanks for your contributions.

I would make two modifications:
- remove the `sortable-group-id="<?php echo $item->catid; ?>"` to allow to sort across multiple categories. We are talking here about feature articles that can be across categories.
- In the list filter xml i would put the fp.ordering ordering options in first place because it's the first column. That would normalize with the other table list view.

File: administrator/components/com_content/views/featured/tmpl/default.php (hunk 2 above)
Comment (andrepereiradasilva) [inline administrator/components/com_content/views/featured/tmpl/default.php]: It's already normalized it with `title ASC` as default ordering.

**Comment by nishadi** [comment]
@andrepereiradasilva
Thanks you for the PRs and clarifications. I have accepted them.

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 75cba61e770f8f59537e628128d704302d36d9e6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by Webdongle** [comment]
I have tested this item :red_circle: unsuccessfully on 75cba61e770f8f59537e628128d704302d36d9e6<br /><br />The last change broke it.  Using Test sample data ... the 'Professionals' article is always at the top.  I have changed the layout settings in the menu item to 0 (zero) leading articles and 1 column.  Wherever the 'Professionals' article is in the (admin)order it always appears at the top in the front end.

@andrepereiradasilva 's addition has broke the patch<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by andrepereiradasilva** [comment]
@Webdongle sorry? my patch only removed the limitation do order between multiple categories, as it should be. Before the patch has limited to order with drag and drop in one category, after the patch you can order with drag and drop between several categories.

See https://github.com/joomla/joomla-cms/pull/9564/commits/ab1d799153cb602719197f9f8971d04ddd9e8738

**Comment by Webdongle** [comment]
@andrepereiradasilva when I tested before your pr was added the articles displayed correctly in the front end.  It could have been the removal of the 'Ordering' column'.  If @nishadi checks by removing each edit and setting the menu item 1 column then that should sow which broke it

Also point of testing etiquette ... although it;s not your patch you have added to it.  Therefore should your test results in the tracker be discounted ?  I mean no disrespect by that question.

**Comment by andrepereiradasilva** [comment]
I have not tested this item.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by andrepereiradasilva** [comment]
> @andrepereiradasilva when I tested before your pr was added the articles displayed correctly in the front end. It could have been the removal of the 'Ordering' column'. If @nishadi checks by removing each edit and setting the menu item 1 column then that should show which broke it

ok, so no problem with my patch them.

> Also point of testing etiquette ... although it;s not your patch you have added to it. Therefore should your test results in the tracker be discounted ? I mean no disrespect by that question.

Actually it makes perfectly sense. If i contributed to the PR i shouldn't test.  Thank you for the warning. I removed my test.
I guess i need some lessons of "testing etiquette" :wink:

**Comment by Webdongle** [comment]
@andrepereiradasilva 
settings 01

settings 02

settings 03

Same result with
settings 04

There is a Category order in the menu item edit screen and imho removing sortable-group-id="<?php echo $item->catid; ?>" messed up the way the Component reads the menu item settings.

Thae Category order Setting in the menu item edit screen may be redundant (I have yet to test) but methinks it is required.  Because the webmaster may wish to display Category order first then other order per category.
e.g.

Cats
--- Moggies
--- Persians

Dogs
--- Alsations
--- Terrier

Therefore removing sortable-group-id="<?php echo $item->catid; ?>" breaks the usability in any case.  And that change should be reverted.

This is not a 'knock' at your efforts. ... they are appreciated.  Just that I test/post in a 'matter of fact' way without emotion.

**Comment by andrepereiradasilva** [comment]
@Webdongle 
Check the your backend column order (you are ordering **by title** there). You should be ordering by order.

image

image

image

**Comment by andrepereiradasilva** [comment]
@Webdongle 
> There is a Category order in the menu item edit screen and imho removing sortable-group-id="<?php echo $item->catid; ?>" messed up the way the Component reads the menu item settings.

That "Category order" option is for, when you have feature articles in several categories, order by that "Category order" first and then by "Article order" (the next field).

You can see that behaviour in https://github.com/joomla/joomla-cms/blob/staging/components/com_content/models/featured.php#L124-L130

My change does not change that. In fact i make the change to allow to use correct "No order" when you have feature articles across categories, like it was before this PR.

**Comment by Webdongle** [comment]
I have tested this item :white_check_mark: successfully on 75cba61e770f8f59537e628128d704302d36d9e6<br /><br />Retested and Full Success<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by Webdongle** [comment]
@andrepereiradasilva 

Yep ... that was it had wrong sort column.  Have deleted my previous test result and added a successful one.  Thanks for finding my PEBKAC

I know what the "Category order" option is for.  Had to have breakfast before I had chance to test.  I changed the category of one of the featured articles ... applying the patch had no adverse affect.

**Comment by andrepereiradasilva** [comment]
@lunalars can you test this again after latest changes so it has to successfully tests.

**Comment by lunalars** [comment]
I have tested this item :white_check_mark: successfully on 75cba61e770f8f59537e628128d704302d36d9e6<br /><br />Tested on staging with test content: changed ordering is saved and articles are displayed accordingly.

(And sorry for being late)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by nishadi** [comment]
Thanks for testing the PR

**Comment by brianteeman** [comment]
RTC - thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9564">issues.joomla.org/joomla-cms/9564</a>.</sub>

**Comment by rdeutz** [comment]
@nishadi cloud you please check the merge conflicts, thanks

**Comment by Kubik-Rubik** [comment]
Thank you for creating this. It’s been some time since you created this and there are now some merge conflicts that prevent a direct merge. To save time I have created a new Pull Request from your code and am closing this here.

Check https://github.com/joomla/joomla-cms/pull/10313
